src/classifier/Pipe.cpp

```diff
@@ -443,15 +457,110 @@ void Pipe::Run() {
   LOG(INFO) << "Number of instances: " << num_instances;
   LOG(INFO) << "Time: " << diff_ms(end,start);
 
+  LOG(INFO) << "Cache size: " << parameters_->caching_weights_.size();
+  LOG(INFO) << "Cache hits: " << parameters_->caching_weights_.hits();
+  LOG(INFO) << "Cache misses: " << parameters_->caching_weights_.misses();
+
   if (options_->evaluate()) EndEvaluation();
 }
 
+//Variant of void Pipe::Run() , that launches threads for processing each instance
+void Pipe::RunWithThreads() {
+
+	timeval start, end;
+	gettimeofday(&start, NULL);
+
+	if (options_->evaluate()) BeginEvaluation();
+
+	reader_->Open(options_->GetTestFilePath());
+	writer_->Open(options_->GetOutputFilePath());
+
+	int num_instances = 0;
+
+	
+	if (instances_.size() != 0){
+		cerr << "Instances vector is not empty at start of Run procedure" << endl;
+		abort();
+	}
+
+	running_multithreaded_ = true;
+
+	Instance *instance = reader_->GetNext();
+	instances_.push_back(instance);
+	while (instance) {
+		//Other instances-related objects 
+		Instance *formatted_instance = GetFormattedInstance(instance);
+		formatted_instances_.push_back(formatted_instance);
+		Instance *output_instance = instance->Copy();
+		output_instances_.push_back(output_instance);
+		//Launch thread
+		list_of_threads_.push_back(std::thread(&Pipe::RunThreaded, this, instance, formatted_instance, output_instance) );
+		
+		//Get next instance
+		instance = reader_->GetNext();
+		instances_.push_back(instance);
+		++num_instances;
+	}
+	//for every thread launched
+	for (int i = 0; i < list_of_threads_.size(); i++) {
+		//wait for that thread
+		list_of_threads_[i].join();
+		//process its output data
+		writer_->Write(output_instances_[i] );
+		//discard memory objects
+		if (formatted_instances_[i] != instances_[i] ){
+			delete formatted_instances_[i];
+		}
+		delete instances_[i];
+		delete output_instances_[i];
+	}
+	// clear std::vectors
+	list_of_threads_.clear();
+	instances_.clear();
+	formatted_instances_.clear();
+	output_instances_.clear();
+	//close channels
+	writer_->Close();
+	reader_->Close();	
+	//Logging
+	gettimeofday(&end, NULL);
+	LOG(INFO) << "Number of instances: " << num_instances;
+	LOG(INFO) << "Time: " << diff_ms(end, start);
+
+	if (options_->evaluate()) EndEvaluation();
+}
+
+void Pipe::RunThreaded(Instance *instance, Instance * formatted_instance, Instance *output_instance) {
```

Comment:
Also, the name of this function looks wrong. It seems more general than threads. Maybe rename it: RunSingleInstance? And shouldn't the instance be formatted inside (in which case the second argument would be eliminated)?